pkg/server/stream/v3/subscription.go

```diff
@@ -0,0 +1,176 @@
+	explicitWildcard = "*"
+)
+
+// Subscription stores the server view of a given type subscription in a stream.
+type Subscription struct {
+	// wildcard indicates if the subscription currently has a wildcard watch.
+	wildcard bool
+
+	// allowLegacyWildcard indicates that the stream never provided any resource
+	// and is de facto wildcard.
+	// As soon as a resource or an explicit subscription to wildcard is provided,
+	// this flag will be set to false
+	allowLegacyWildcard bool
+
+	// subscribedResourceNames provides the resources explicitly requested by the client
+	// This list might be non-empty even when set as wildcard.
+	subscribedResourceNames map[string]struct{}
+
+	// returnedResources contains the resources acknowledged by the client and the acknowledged versions.
+	returnedResources map[string]string
+}
+
+// newSubscription initializes a subscription state.
+func newSubscription(wildcard bool, initialResourceVersions map[string]string) Subscription {
+	state := Subscription{
+		wildcard:                wildcard,
+		allowLegacyWildcard:     wildcard,
+		subscribedResourceNames: map[string]struct{}{},
+		returnedResources:       initialResourceVersions,
+	}
+
+	if initialResourceVersions == nil {
+		state.returnedResources = make(map[string]string)
+	}
+
+	return state
+}
+
+func NewSotwSubscription(subscribed []string) Subscription {
+	sub := newSubscription(len(subscribed) == 0, nil)
+	sub.SetResourceSubscription(subscribed)
+	return sub
+}
+
+// SetResourceSubscription updates the subscribed resources (including the wildcard state)
+// based on the full state of subscribed resources provided in the request
+// Used in sotw subscriptions
+// Behavior is based on
+// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return
+func (s *Subscription) SetResourceSubscription(subscribed []string) {
+	if s.allowLegacyWildcard {
+		if len(subscribed) == 0 {
+			// We were wildcard based on legacy behavior and still don't request any resource
+			// The watch remains wildcard
+			return
+		}
+
+		// A resource was provided (might be an explicit wildcard)
+		// Documentation states that we should no longer allow to fallback to the previous case
+		// and no longer setting wildcard would no longer subscribe to anything
+		s.allowLegacyWildcard = false
+	}
+
+	subscribedResources := make(map[string]struct{}, len(subscribed))
+	explicitWildcardSet := false
+	for _, resource := range subscribed {
+		if resource == explicitWildcard {
+			explicitWildcardSet = true
+		} else {
+			subscribedResources[resource] = struct{}{}
+		}
+	}
+
+	// Explicit subscription to wildcard as we are not in legacy wildcard behavior
+	s.wildcard = explicitWildcardSet
+	s.subscribedResourceNames = subscribedResources
+}
+
+func NewDeltaSubscription(subscribed, unsubscribed []string, initialResourceVersions map[string]string) Subscription {
+	sub := newSubscription(len(subscribed) == 0, initialResourceVersions)
```

Comment:
Similarly, isn't it possible to set an explicit wildcard in the first delta request?